### DISCUSSION TITLE
ChatGPT with PyVista

### DISCUSSION THREAD

**Discussion by MatthewFlamm** [pyvista/pyvista#3984] [Show and tell]
Just for fun, I asked ChatGPT for "Python code for plotting red torus using PyVista", which it completes accurately:

```py
import pyvista as pv
import numpy as np

# Define the parameters of the torus
R = 1.0 # radius of the tube
r = 0.25 # radius of the torus
N = 50

# Generate the torus surface
theta = np.linspace(0, 2 * np.pi, N)
phi = np.linspace(0, 2 * np.pi, N)
theta, phi = np.meshgrid(theta, phi)
x = (R + r * np.cos(phi)) * np.cos(theta)
y = (R + r * np.cos(phi)) * np.sin(theta)
z = r * np.sin(phi)
torus = pv.StructuredGrid(x, y, z)

# Plot the torus and set its color to red
p = pv.Plotter()
p.add_mesh(torus, color="red")
p.show()
```
![image](https://user-images.githubusercontent.com/39341281/217656256-b39c6c1c-fc05-4e12-b7e4-e89836f4c93d.png)

Additionally asking it to "add a blue cylinder inside torus" fails. It interestingly misses that there is a built-in cylinder command and completely messes up the simple cylinder formula.

```py
import pyvista as pv
import numpy as np

# Define the parameters of the torus
R = 1.0 # radius of the tube
r = 0.25 # radius of the torus
N = 50

# Generate the torus surface
theta = np.linspace(0, 2 * np.pi, N)
phi = np.linspace(0, 2 * np.pi, N)
theta, phi = np.meshgrid(theta, phi)
x = (R + r * np.cos(phi)) * np.cos(theta)
y = (R + r * np.cos(phi)) * np.sin(theta)
z = r * np.sin(phi)
torus = pv.StructuredGrid(x, y, z)

# Generate the cylinder surface
cyl_x = np.linspace(-R, R, N)
cyl_y = np.linspace(-R, R, N)
cyl_z = np.zeros(N)
cyl_z = np.tile(cyl_z, (N, 1))
cyl_x, cyl_y = np.meshgrid(cyl_x, cyl_y)
cylinder = pv.StructuredGrid(cyl_x, cyl_y, cyl_z)

# Plot the torus and cylinder and set their colors
p = pv.Plotter()
p.add_mesh(torus, color="red")
p.add_mesh(cylinder, color="blue")
p.show()
```

![image](https://user-images.githubusercontent.com/39341281/217656035-ed06c303-dab0-4467-a2fc-f9a8b06e9804.png)

If you originally ask it to "python code for plotting blue cylinder in pyvista" , it creates the cylinder correctly this time, but then slightly messes up the plotting command.

```py
import pyvista as pv

cylinder = pv.Cylinder(height=1.0, radius=0.5, center=(0.0, 0.0, 0.0))
cylinder.color = "blue"
cylinder.plot()
```

By asking "python code for a movie of a small red sphere circling around a larger blue sphere in pyvista" at first it mixes up pyvista and matplotlib commands.  It also uses the wrong syntax for updating the position of the red sphere.  I tried to guide it to working code, but I had to more or less write the code myself this time (not shown).

```py
import pyvista as pv
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as animation

# Create the blue sphere
blue_sphere = pv.Sphere(radius=1.0)

# Create the red sphere
red_sphere = pv.Sphere(center=(1.5, 0, 0), radius=0.3)

# Define the animation function
def animate(i):
    angle = np.pi * 2 * i / 360
    x = 1.5 * np.cos(angle)
    y = 1.5 * np.sin(angle)
    red_sphere.center = (x, y, 0)
    plotter = pv.Plotter()
    plotter.add_mesh(blue_sphere, color='blue')
    plotter.add_mesh(red_sphere, color='red')
    plotter.show()

# Create the animation
ani = animation.FuncAnimation(plt.gcf(), animate, frames=360)

# Save the animation as a movie file
ani.save('movie.mp4', writer='ffmpeg')
```

Not surprising that the quality of the result depends on the complexity of the ask.  PyVista code is probably not heavily represented in its training data, yet it succeeds occasionally.  It often uses plausible but nonexistent functionality like `red_sphere.center = (x, y, 0)` in the last example.  Hopefully it doesn't trawl the data here to reinforce it's ~bad~ underrepresented training.

**Comment by banesullivan** [comment]
This is cool, thanks for posting!! @akaszynski might enjoy this -- Alex actually opened a non-trivial PR a while back that was (I think) 99% written by ChatGPT... I don't know if he made that known. I forget which PR exactly, so I'll let Alex link it

**Reply by akaszynski** [reply]
I indeed was able to get it to make almost all of https://github.com/pyvista/pyvista/pull/3606. In fact, we have a repo that does this (poorly, but it does it) at https://github.com/pyvista/pyvista-bot and specifically with https://github.com/pyvista/pyvista-bot/blob/main/single_shot_example.py

This is effectively "one shot" training. I give it an example vtk class definition scraped from the oxygen docs and the corresponding pyvista function. It's quite good, and in a separate function I was able to get it to write the docstring (which it great at). Unit testing wasn't as easy, but with better training data we'd be able to get it to write better tests.

**Reply by MatthewFlamm** [reply]
Wow this is great!  I did not know about the training aspect.  I have also found that using it to generate docstrings is the best use so far.  This is where getting close, but not exact, is really helpful.

**Comment by banesullivan** [comment]
I wonder if we could use this to improve our docs experience: https://github.com/arc53/docsgpt

I’m imagining a PyVista version of “Clippy”


![](https://cdn.vox-cdn.com/thumbor/PoaGaIZ1cYLqlLwelsO4_ncBdhs=/0x0:1062x705/1400x1050/filters:focal(531x353:532x354)/cdn.vox-cdn.com/uploads/chorus_asset/file/15979182/clippy.jpg)
